### ISSUE TITLE
Can't install mods

### ISSUE THREAD

**Issue by cleveHEX** [vcmi/vcmi#2731]
**Describe the bug**
Every time when I try to download and install a mod, it shows this:
![obrazek](https://github.com/vcmi/vcmi/assets/9021297/f8c5a012-cee3-4b7e-883b-ce3f732d9a9b)

**Game logs**
[VCMI_Launcher_log.txt](https://github.com/vcmi/vcmi/files/12503700/VCMI_Launcher_log.txt)

**To Reproduce**
Steps to reproduce the behaviour:
1. Install vcmi via dnf
2. Open VCMI
3. I selected Czech language and it wanted to download some other packages, it failed.
4. Restarted PC, tried opening the launcher and downloading manually and this happens.

**Expected behavior**
Mods install.

**Actual behavior**
Mods fail to install

**Did it work earlier?**
It's like this since I installed it.

**Screenshots**
Above

**Version**
 - OS: Fedora 38
 - Version: 1.3.1

**Additional context**

N/A

**Comment by Nordsoft91** [comment]
in VCMI data folder, find and delete "downloads" folder then try again.

**Comment by cewbdex** [comment]
I have deleted ~/.cache/vcmi/downloads but no change.

**Comment by Nordsoft91** [comment]
it's required more information to analyse a bug
1. which mods do you have installed (should be visible on mods screen in launcher)?
2. can you check that your use default repository? Settings tab -> Mod repositories -> Default repository (check)
3. On screenshot you cannot install vcmi-extras, can you search for `vcmi-extras.zip` and attach it?

**Comment by cewbdex** [comment]
1. Only the base files are installed, no mod was successful to install. I tried Czech language, vcmi-extras and also one mod that VCMI wanted to install itself after first run, I don't remember the name now.
2. ![obrazek](https://github.com/vcmi/vcmi/assets/9021297/5402ddb9-70c6-44db-b0c1-e912ee6b8b96)
3. I can't locate the file, could you tell me, where it should be located?
```
[cleve@pk-nova ~]$ find / -name vcmi-extras*
/home/cleve/.cache/vcmi/downloads/vcmi-extras.json
```

**Comment by misiokles** [comment]
There were silimiar problem about week ago (from Chech also). Problem was with internet provider that blocked access to GitHub.

**Comment by cewbdex** [comment]
I don't think that is the problem here, as it downloads the index properly and now I have tested it with different provider and the same result. I can see the download progress too. I don't know how to install manually, is extracting to some folder enough? Is there an option to have more detailed logging?

**Comment by misiokles** [comment]
Download links for each mods are given here:
https://github.com/vcmi/vcmi-mods-repository/blob/develop/vcmi-1.3.json

Just copy/paste in browser link to .zip files.

After unpacking copy folders to /Documents/My Games/vcmi/Mods

Proper folder structure should be /Mods/mod_name/mod.json 
Properly manually installed mod should appear in launcher.

**Comment by IvanSavenko** [comment]
Hi. Can you check contents of `/home/cleve/.cache/vcmi/downloads/` directory?
Does it contains any zip archives? (note that filename won't be human-readable)

Unfortunately we don't have any detailed logging here, so not sure how to investigate this further.

You can install mods manually using misiokles link.
On Linux path to mods directory would be `~/.local/share/vcmi/Mods/`.
After extraction, you need to have directory structure like `~/.local/share/vcmi/Mods/(mod name)/mod.json`.

**Comment by cewbdex** [comment]
I have [recorded the issue](https://youtu.be/SZx8O0DBlEg) for you.
Only JSON and PNG files are present in `/home/cleve/.cache/vcmi/downloads/`. I have also searched my whole file system and they are not anywhere. However, I tried to download HOTA, as I thought it will be big and I looked into the folder and there was a file `hota.zip` (or something like this, it was quick), but the moment the error window appeared it disappeared.
I will try the manual method now.

**Comment by cewbdex** [comment]
So I pasted the files like this:
![obrazek](https://github.com/vcmi/vcmi/assets/9021297/3c260539-7d30-44ec-86e4-8e371f8255d7)
And VCMI launcher does not load it, log is here:
[VCMI_Launcher_log.txt](https://github.com/vcmi/vcmi/files/12507409/VCMI_Launcher_log.txt)

**Comment by cewbdex** [comment]
I also tried to merge the `Mods` folder from `vcmi-extras-vcmi-1.3` to the parent `Mods` directory like this:
![obrazek](https://github.com/vcmi/vcmi/assets/9021297/490ed03f-3445-4a8d-a25a-f60346491e09)

And now it works for me like this:
![obrazek](https://github.com/vcmi/vcmi/assets/9021297/32b0701e-f0a6-4b25-beb5-ecc49377e663)
Maybe some problem with the "pack" json? I don't know

**Comment by Nordsoft91** [comment]
when you downloaded mod, zip should appear in downloads folder. When error occurs it probably could still be there. If possible to catch it, can you attach to the issue?

Also, maybe you have one more downloads folder in another location? .local/share/vcmi ?

**Comment by cewbdex** [comment]
The file is removed the moment when the error message appears, so that's when the file finishes downloading, I don't see a way how to catch the file before it gets deleted by vcmi. If you know, tell me.

No, vcmi seems to nicely respect the folder structure of my Linux based system, actually tell gj to whoever did it, as many apps ignore it and have their own folder in home. @IvanSavenko pointed me well.

**Comment by Nordsoft91** [comment]
Can you try following?
1. download archive by direct link (e.g. `"https://github.com/vcmi-mods/vcmi-extras/archive/refs/heads/vcmi-1.3.zip`) 
2. verify that it's not corrupted
4. copy zip to downloads folder without unpacking
5. verify that such mod is not installed (folder vcmi-extras is not present in mods folder)
6. launch vcmi - it should detect zip and start to extract it

**Comment by Nordsoft91** [comment]
> And VCMI launcher does not load it, log is here:

I was able to reproduce. To fix it locally you need to rename folder from "vcmi-extras-vcmi-1.3" to "vcmi-extras"

**Comment by Nordsoft91** [comment]
can be permissions problem? does vcmi have a write access to Mods folder?

**Comment by cewbdex** [comment]
Will try. Is that a bug or it's not supposed to load directories with numbers?

I will check, but it's just a casual folder in user directory, it should have, I would compare it to Windows' appdata.

**Comment by IvanSavenko** [comment]
> Will try. Is that a bug or it's not supposed to load directories with numbers?

Kind of both? Mod directory is often used to reference mods, so renaming mod directory might lead to issues. So yes, try renaming it to "vcmi-extras".

It is however strange that Launcher failed to load mod from "vcmi-extras-vcmi-1.3" folder.

> I will check, but it's just a casual folder in user directory, it should have, I would compare it to Windows' appdata.

Yeah, probably unrelated to permissions. If it was then vcmi won't be able to save json's and png's in there. Sounds like a bug with vcmi but unfortunately no idea how to investigate it properly.

**Comment by cewbdex** [comment]
> Kind of both? Mod directory is often used to reference mods, so renaming mod directory might lead to issues. So yes, try renaming it to "vcmi-extras".

This did the job.

**Comment by Nordsoft91** [comment]
@cewbdex still make sense to try `chmod a+x ~/.local/share/vcmi/Mods`

**Comment by IvanSavenko** [comment]
Won't change anything.
That would have produced different error message.

"Mod archive is invalid or corrupted" means that vcmi wasn't able to locate mod.json in zip

**Comment by tmysik** [comment]
The VCMI team could perhaps prepare a new version (or better, a custom build specifically for this issue?) which does not remove the downloaded mod file? This functionality could be "hidden" behind some system property or environment variable, perhaps? Or, simply more verbose logging for this particular situation, if it is possible. Just an idea, sorry if there is a better way what to do now 🤔

**Comment by jirkafm** [comment]
I have same issue on my laptop (FC38), the following warning messages were in the bottom of ~/.cache/vcmi/VCMI_Launcher_log.txt when I tried to install In the Wake of Gods mod (mod doesn't matter, nothing can be installed in my case)

```
WARN mod [7f134e7182c0] - File <unknown> is not a valid JSON file!
WARN mod [7f134e7182c0] - At line 15, position 1 warning: Comma expected
```

**Comment by jirkafm** [comment]
> Download links for each mods are given here: https://github.com/vcmi/vcmi-mods-repository/blob/develop/vcmi-1.3.json
> 
> Just copy/paste in browser link to .zip files.
> 
> After unpacking, copy folders to /Documents/My Games/vcmi/Mods
> 
> Proper folder structure should be /Mods/mod_name/mod.json Properly manually installed mod should appear in launcher.

I can confirm, that manual installation works properly.

**Comment by IvanSavenko** [comment]
@jirkafm what's your OS / distribution? Also Fedora?

**Comment by khardix** [comment]
I'm running into the same issue, also on Fedora 38.

I was suspecting selinux (which is the usual suspect on Fedora), but disabling it made no difference – the download fails either way.

**Comment by jirkafm** [comment]
> @jirkafm what's your OS / distribution? Also Fedora?

Yes, I slightly mentioned it in previous post (FC38 - Fedora Core 38). Sorry, I didn't realize it's not so obvious.

**Comment by IvanSavenko** [comment]
Don't see anything wrong with this part of our code unfortunately...

Checked Fedora packaging for vcmi from here: https://github.com/rpmfusion/vcmi
- don't see anything unusual
- Fedora 39 uses Qt6 which we haven't tested, but reports are for 38
- In theory could be caused by version of minizip that we haven't tested, but that won't explain why manual install works

Will add better logging here for 1.3.2 - looks like all errors are currently silently ignored here.

**Comment by khardix** [comment]
> * In theory could be caused by version of minizip that we haven't tested, but that won't explain why manual install works

Well, from the logs (see [comment above](https://github.com/vcmi/vcmi/issues/2731#issuecomment-1712829344)) it seems that the unzipped mod.json is considered invalid – perhaps unzipping generates some weird characters that confuse the parser?

Manually unzipping the mod would sidestep that issue…

**Comment by IvanSavenko** [comment]
Ah, yes. Yeah, probably somehow related to minizip. Will try to look into it.

For some reason forgot that mod install is largely the only part of game that uses zip's. Another one is maps that use vcmi-specific format, but those are very few.

**Comment by runlevel5** [comment]
@IvanSavenko Hello, I am currently the maintainer of RPMFusion's RPM package of VCMI. I can confirm it is _definitely_ due to minizip-ng that RPMFusion's Fedora vcmi package uses. While waiting for vcmi to support minizip-ng, I will make changes on my side to use bundled minizip instead

**Comment by GeorgeK1ng** [comment]
Should be fixed in 1.7 develop
